arcgis-ios-sdk-samples/Edit data/Create and save KML file/CreateAndSaveKMLViewController.swift

```diff
@@ -0,0 +1,270 @@
+    // Prompt feature selection action sheet.
+    @IBAction func addFeature() {
+        let alertController = UIAlertController(title: "Select Feature", message: nil, preferredStyle: .actionSheet)
+        let pointAction = UIAlertAction(title: "Point", style: .default) { (_) in
+            self.addPoint()
+        }
+        alertController.addAction(pointAction)
+        let polylineAction = UIAlertAction(title: "Polyline", style: .default) { (_) in
+            self.addPolyline()
+        }
+        alertController.addAction(polylineAction)
+        let polygonAction = UIAlertAction(title: "Polygon", style: .default) { (_) in
+            self.addPolygon()
+        }
+        alertController.addAction(polygonAction)
+        
+        // Add "cancel" item.
+        let cancelAction = UIAlertAction(title: "Cancel", style: .cancel, handler: nil)
+        alertController.addAction(cancelAction)
+        
+        alertController.popoverPresentationController?.barButtonItem = addButton
+        present(alertController, animated: true, completion: nil)
+    }
+    
+    // Prompt options to allow the user to save the KMZ file.
+    @IBAction func saveKMZ(_ sender: UIBarButtonItem) {
+        let kmzProvider = KMZProvider(document: kmlDocument)
+        let activityViewController = UIActivityViewController(activityItems: [kmzProvider], applicationActivities: nil)
+        activityViewController.popoverPresentationController?.barButtonItem = sender
+        present(activityViewController, animated: true)
+        activityViewController.completionWithItemsHandler = { (activityType: UIActivity.ActivityType?, completed: Bool, arrayReturnedItems: [Any]?, error: Error?) in
+            kmzProvider.deleteKMZ()
+        }
+    }
+
+    // Complete the current sketch and add it to the KML document.
+    @IBAction func completeSketch() {
+        let geometry = mapView.sketchEditor?.geometry
+        let projectedGeometry = AGSGeometryEngine.projectGeometry(geometry!, to: spatialRef)
+        let kmlGeometry = AGSKMLGeometry(geometry: projectedGeometry!, altitudeMode: .clampToGround)
+        currentPlacemark = AGSKMLPlacemark(geometry: kmlGeometry!)
+        currentPlacemark!.style = kmlStyle
+        kmlDocument.addChildNode(currentPlacemark!)
+        mapView.sketchEditor?.stop()
+        updateToolbarItems()
+        actionButtonItem?.isEnabled = true
+    }
+    
+    // Reset the KML.
+    @IBAction func resetKML() {
+        mapView.map?.operationalLayers.removeAllObjects()
+        currentPlacemark = nil
+        kmlDocument = AGSKMLDocument()
+        let kmlDataset = AGSKMLDataset(rootNode: kmlDocument)
+        kmlLayer = AGSKMLLayer(kmlDataset: kmlDataset)
+        mapView.map?.operationalLayers.add(kmlLayer!)
+    }
+    
+    var sketchCreationMode: AGSSketchCreationMode?
+    var kmlDocument = AGSKMLDocument()
+    let spatialRef = AGSSpatialReference.wgs84()
+    var kmlStyle = AGSKMLStyle()
+    var currentPlacemark: AGSKMLPlacemark?
```

Comment:
Correct me if I'm wrong, but I don't think this property is needed anymore. It is being assigned to in `completeSketch()` and set to `nil` in `resetKML()`, but it isn't read anywhere, so that doesn't seem to be necessary.

Reply:
`currentPlacemark` is read in `completeSketch()` in order to add a child node but I can take it out in `resetKML()`. On a related note, this topic lead me to realize possible problems when `resetKML()` is called during sketch editing so I've decided to disable the button while sketch mode is activated.